lib/gemstash/s3_storage.rb

```diff
@@ -0,0 +1,290 @@
+      def initialize(folder, version)
+        super("Gemstash storage version #{Gemstash::S3::STORAGE_VERSION} does " \
+               "not support version #{version} found at #{folder}")
+      end
+    end
+
+    def initialize(folder, root: true)
+      @folder = folder
+      check_storage_version if root
+      @client = Aws::S3::Client.new(
+        access_key_id: gemstash_env.config[:aws_access_key_id],
+        secret_access_key: gemstash_env.config[:aws_secret_access_key],
+        region: gemstash_env.config[:region]
+      )
+      @bucket_name = gemstash_env.config[:bucket_name]
+      @object_name = @folder
+    end
+
+    def check_credentials?
+      @client.get_bucket_location(bucket: @bucket_name).location_constraint == gemstash_env.config[:region]
+    end
+
+    def delete_with_prefix(prefix = @folder)
+      Aws::S3::Resource.new(client: @client).bucket(@bucket_name).objects(prefix: prefix).batch_delete!
+    end
+
+    def resource(id)
+      S3Resource.new(@folder, id, @client, @bucket_name)
+    end
+
+    def for(child)
+      S3.new(File.join(@folder, child), root: false)
+    end
+
+    def self.for(name)
+      new(File.join(gemstash_env.config[:s3_path], name))
+    end
+
+    def self.metadata
+      file = gemstash_env.base_file("metadata.yml")
+      unless File.exist?(file)
+        gemstash_env.atomic_write(file) do |f|
+          f.write({ storage_version: Gemstash::S3::STORAGE_VERSION,
+                    gemstash_version: Gemstash::VERSION }.to_yaml)
+        end
+      end
+
+      YAML.load_file(file)
+    end
+
+  private
+
+    def check_storage_version
+      version = Gemstash::S3.metadata[:storage_version]
+      return if version <= Gemstash::S3::STORAGE_VERSION
+
+      raise Gemstash::S3::VersionTooNew.new(@folder, version)
+    end
+  end
+
+  # A resource within the storage engine. The resource may have 1 or more files
+  # associated with it along with a metadata Hash that is stored in a YAML file.
+  class S3Resource
+    include Gemstash::Env::Helper
+    include Gemstash::Logging
+    attr_reader :name, :folder, :client
+    VERSION = 1
+
+    # If the storage engine detects a resource was originally saved from a newer
+    # version, this error is thrown.
+    class VersionTooNew < StandardError
+      def initialize(name, folder, version)
+        super("Gemstash resource version #{Gemstash::S3Resource::VERSION} does " \
+               "not support version #{version} for resource #{name.inspect} " \
+               "found at #{folder}")
+      end
+    end
+    def initialize(folder, name, client, bucket_name)
+      @folder = folder
+      @name = name
+      safe_name = sanitize(@name)
+      digest = Digest::MD5.hexdigest(@name)
+      child_folder = "#{safe_name}-#{digest}"
+      @folder = File.join(@folder, child_folder)
+      @client = client
+      @s3resource = Aws::S3::Resource.new(client: @client).bucket(bucket_name)
+      @properties = nil
```

Comment:
Don't need to initialize this as nil

Reply:
Also, why do you think so? Isn't it important to set the properties to nil when it is first initialized?

Reply:
If you don't initialize it it will always be nil

Reply:
Fixed in commit #859fdb8